### PR TITLE
feat: Detect Gateway API version

### PR DESCRIPTION
**Pull request by horjulf**
#### Added

Detects the Gateway API version available for each kind in the `route` object.

### Applicable issues

- fixes #

## Additional information

https://github.com/kubernetes-sigs/gateway-api/releases/tag/v0.8.0

Version `0.8.0` of Gateway API drops `v1alpha2` version for `Gateway`, `GatewayClass`, and `HTTPRoute`.

## Checklist

- [x] Title of the PR conforms to the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard.
- [x] Scope of the of the PR title contains the chart name.
- [x] Chart version in `Chart.yaml` has been bumped according to [Semantic Versioning](https://semver.org/).
- [x] Chart `artifacthub.io/changes` changelog annotation has been updated in `Chart.yaml`. See [Artifact Hub documentation](https://artifacthub.io/docs/topics/annotations/helm/#supported-annotations) for more info.
- [x] Variables have been documented in the `values.yaml` file.

### CODE REVIEW

**Comment by horjulf** [comment]
@bjw-s should I rebase this into the `rewrite` branch instead ?

WDYT of updating the default to `v1beta1` version ? The TCP/UDP along with some other types still use `v1alpha2`, so its likely useful to keep it configurable.

**Comment by bjw-s** [comment]
I'm going to be merging the `rewrite` branch into `development` tonight or tomorrow. Then I will use `development` as a staging area to update docs etc before the initial release of 2.0.

**Comment by horjulf** [comment]
@bjw-s thanks, rebased, should we take the opportunity of the major bump and change the default API version to `v1beta1` ?

**Comment by bjw-s** [comment]
> @bjw-s thanks, rebased, should we take the opportunity of the major bump and change the default API version to `v1beta1` ?

I'm definitely not against bumping the version. But since I don't use the Gateway API yet I have no idea about any other potential implications of the change, which is why I'm a bit hesitant to make it configurable through the values.yaml

**Comment by horjulf** [comment]
> I'm definitely not against bumping the version. But since I don't use the Gateway API yet I have no idea about any other potential implications of the change, which is why I'm a bit hesitant to make it configurable through the values.yaml

We could bump it directly, however it would break the non-http types (routeKind) since they are only available in `v1alpha2`.

Basically the >= `0.8.0` spec dropped `HTTPRoute` from `v1alpha2` and is now exclusively available in `v1beta1`.

Spec at https://gateway-api.sigs.k8s.io/references/spec/#api-specification

Depending on what version of Gateway API CRDs someone is using, they will need different API versions to get it working.

Alternatively we could try to check capabilities and set the version automatically, WDYT ?

**Comment by horjulf** [comment]
> Alternatively we could try to check capabilities and set the version automatically

Something like:
```
  {{- $routeKind := $routeObject.kind | default "HTTPRoute" -}}
  {{- $apiVersion := "gateway.networking.k8s.io/v1alpha2" -}}
  {{- if $rootContext.Capabilities.APIVersions.Has (printf "gateway.networking.k8s.io/v1beta1/%s" $routeKind) }}
    {{- $apiVersion = "gateway.networking.k8s.io/v1beta1" -}}
  {{- end -}}
```

**Comment by bjw-s** [comment]
I'd be cool with a solution like that, that would keep it relatively clean. Hopefully those APIs reach v1 relatively soon to get all the versions aligned again.

**Comment by horjulf** [comment]
@bjw-s updated to check capabilities.

File: charts/library/common/Chart.yaml [hunk 1]
```diff
@@ -25,3 +25,5 @@ annotations:
       description: Added support for ordering (init)Containers within a controller.
     - kind: added
       description: Added support for creating Network Policies.
+    - kind: feat
+      description: Detect `gateway.networking.k8s.io` API version
```
Comment (bjw-s) [inline charts/library/common/Chart.yaml:29]:
```suggestion
    - kind: changed
      description: Automatically detect `gateway.networking.k8s.io` API version
```

**Comment by bjw-s** [comment]
Thanks for this, I've merged in to `development` so it should be in the next beta soon

**Pull request merged**